### ISSUE TITLE
CMS <> VEDA Integration Exploration

### ISSUE THREAD

**Issue by j08lue** [NASA-IMPACT/veda-ui#1230]
### Context

We know that making it easier to manage content will make life easier for people managing their VEDA instances. We also assume that the content management needs are neither a) heavy enough or b) unique enough to warrant the development of our own CMS technology. So, we believe it would be valuable to integrate with existing tools. 

We would to develop an informed opinion on our preferred CMS tools to integrate with, and at least have an understanding of how each integration might be implemented. Known tools that we are interested in:
- Strappi or similar - like a Git flow UI
- Drupal (GHG Center)
- WordPress (Earth.gov)

Regarding Drupal:
Many NASA Earth science websites are based on Drupal to make it easy for a team of domain expert contributors and comms specialists to manage content with the help of graphical user interfaces and the usual quality of life features a CMS provides. We know at least some instances, like GHG, might move towards a Drupal implementation so we want to provide guidance on that integration, as well as understand if we can make any adjustments to the core product to better facilitate this integration in the future.

Previous quarterly objective with conversations and links to materials: https://github.com/NASA-IMPACT/veda-architecture/issues/463

## Acceptance criteria

- [ ] We have documented potential implementation paths and concerns for Drupal and Strappi integrations, as well as made a formal recommendation
- [ ] We have shared this information with stakeholders for feedback

**Comment by j08lue** [comment]
The trajectory we are on with our current VEDA UI refactor is that VEDA UI turns into a React component library plus recommendations and examples (template instance) of how to put them together and feed them with data (dataset information, stories, header/footer/nav, page content).

Is it possible to feed these React components we are currently building with data from a (headless) Drupal CMS? Or would they need to work differently?

In my (superficial) understanding, Drupal CMS UI generates content of some form and stores it in an (SQL) database. Then some software (normally PHP) picks it up from the database and renders it into HTML using some website template.

Based on this, the questions are whether it possible that

1. Drupal UI generates the kind of content that VEDA UI components need?
2. A React app which includes VEDA UI components fetches content from the Drupal-managed content database?
3. This setup actually improves the content editing workflow for key contributors - in particular tightening the loop from content editing to rendering (previewing)?

**Comment by aboydnw** [comment]
@freitagb @j08lue just a heads up that we are bringing this initial exploration into the current sprint. We might want to find space to discuss live in the coming weeks

**Comment by j08lue** [comment]
I am really curious to see a record of the possible outcome of relying on headless WordPress or Drupal for content management and VEDA UI for the frontend.

## Editing user experience

1. What types of content can be managed in a headless CMS? 
   1. Stories - prose embedded images etc., but what about our custom MDX blocks like scrollytelling or data visualization?
   2. Pages - how does that play with routing in the frontend? How to configure dynamic features like carousels or card galleries?
   3. Dataset definitions - the dataset overview part can live in the CMS, but perhaps not the (YAML) config?
5. Can we achieve the close feedback loop between content editing and rendered preview that is currently missing?

## Website quality

1. Can we make content fluidly representable throughout the website? Like a story including cards that link to related stories, tools or datasets?
2. Can we achieve fast rendering speed without CMS database reads?

## Development and maintenance effort

1. Would we need to develop custom plugins to WordPress or Drupal (in PHP)?
2. Is additional development required for VEDA UI to make components compatible with getting fed with data from the headless CMS?
3. How much effort is it to build an instance application that works with CMS content?
   1. Where do we bring together technical information (e.g. dataset configuration) with renderable content?

**Comment by snmln** [comment]
@jonas, yes this aligns mostly with what my research is concerning. I need to explore headless Wordpress as an option the majority of my research thus far has been around Drupal and Strapi. Based on the conversation that Anthony and I had yesterday, with Hanbyul and Abdelhak I will expand the area to try and encapsulate a handful of CMS's to reduce any architectural gaps. As for your bullet points I can answer a few currently, the others there need to some additional research and/or architecture conversations and decisions to answer:

> Stories - prose embedded images etc., but what about our custom MDX blocks like scrollytelling or data visualization?
Both drupal and strapi can accommodate this need. The custom MDX blocks might need some additional work to translate or migrate into the CMS's preferred format. 
> Pages - how does that play with routing in the frontend? How to configure dynamic features like carousels or card galleries?
This will ultimately need to take shape as setting a page link and or a dynamic content type in the component element (card, carousel etc.) 
> Dataset definitions - the dataset overview part can live in the CMS, but perhaps not the (YAML) config?
This item I do not have a clear answer for at this time but need to do some additional research around. Hanbyul, did call out in our meeting that the backend team has made some updates/changes in recent time that we might only need to make a referential call to the dataset. 
> Can we achieve the close feedback loop between content editing and rendered preview that is currently missing?
This can be achieved by a preview content functionality. Neither Drupal or Strapi have an out of the box solution for a headless architecture. Both do have documentation and modules that can be leveraged to accomplish this. 
> Can we make content fluidly representable throughout the website? Like a story including cards that link to related stories, tools or datasets?
Both Drupal and Strapi have the ability to create different content types that can be intertwined and reused across the site as a whole. 
> Would we need to develop custom plugins to WordPress or Drupal (in PHP)?
Potentially. Both Drupal and Strapi (I am unsure about Wordpress at this time) have a large community that have already developed a large selection of plugins that we can leverage. But depending on the need there might not be a plugin that exists, which could result in the need of custom development.

**Comment by freitagb** [comment]
Following up with takeaways from some conversations i had at AGU. As a general comment - there continues to be great interest in having a CMS as a core capability for VEDA so we should continue to prioritize this work with some demoable capabilities by the end of Q2. Some key points:

1. Promotion of content to higher level NASA sites (e.g. science.nasa.gov, nasa.gov) is prioritized for content created using Wordpress
 
2. Program managers would also like the ability to edit content on the website. Whichever CMS is chosen should be very intuitive as the audience will extend to people with limited technical or hands-on experience in content editing software.
 
3. Visual editing is a top priority for CMS capabilities. Everyone I talked to asked for a WYSIWYG interface that doesn't require rendering to preview content. I believe when Strapi was first demoed the site needed to be re-rendered to actually see user edits in a web preview.
 
4. Existing storytelling capabilities fall just behind visual editing capabilities. In other words, initial demo could/should include visual editing capabilities even if scrollytelling components aren't fully available within the CMS. That isn't to say we shouldn't get to a place where we can support every feature within a data story, but it's not required for an initial demo to the various implementation teams.

**Comment by aboydnw** [comment]
Thanks for the context @freitagb a couple questions from my end:

> Promotion of content to higher level NASA sites (e.g. science.nasa.gov, nasa.gov) is prioritized for content created using Wordpress.

What does this mean, exactly? Is there a particular file type that Wordpress exports that is easier for these other sites to ingest? Or something else?

> Everyone I talked to asked for a WYSIWYG interface that doesn't require rendering to preview content.

It doesn't surprise me that people would ask for this, because it's something they are familiar with. I don't think it's common to have a WYSIWYG editor like this with a headless CMS, correct me if I'm wrong on that. Assuming we'd need to invest a significant amount of development time, is it worth our time with our size of user pool to do that? 

For me, whether or not we develop a live preview without needing to re-render a site would largely come down to level of effort. If it's easy enough to use some existing tool, then let's do it. I don't think it's worth building our own tool that does this, though, partly because I'm a bit skeptical how often program managers will actually edit content vs handing a content request off to somebody else who is probably savvy enough to just wait for a site to re-render. (this is where the exploration that @snmln is doing comes into play)

**Comment by hanbyul-here** [comment]
I mentioned this briefly during the meeting, but to record - I would like to complement @agurvich 's idea that we can use the hydrating at runtime for preview. ex. We can have `/Preview` route that uses the same component as `/actual-page` component uses, but only pass the data on runtime so we don't have to have separate components to handle the preview


Also in case this helps - These are some of the previous efforts for CMSish system: 
Editor built on client side of VEDA UI app: https://github.com/NASA-IMPACT/veda-ui/pull/604
Netlify CMS: https://github.com/NASA-IMPACT/veda-ui/pull/224

**Comment by j08lue** [comment]
Just for the record, our experimental live-rendering MDX editor is still alive. You can find it on any non-production deployment under `/sandbox/mdxeditor` e.g. https://visex.netlify.app/sandbox/mdxeditor

- ✅ real-time MDX rendering (paste your MDX)
- ✅ support for all blocks (images need to be uploaded / available on URLs already)
- ❌ integration with GitHub for stage-in and stage-out
- ❌ user authentication and authorization
